docs/proposals/201909_bucketfile.md

```diff
@@ -0,0 +1,329 @@
+---
+title: BucketFile for reducing memory usage 
+type: proposal
+menu: proposals
+status: 
+owner: zhulongcheng
+---
+
+### Related Tickets
+
+* https://github.com/thanos-io/thanos/issues/1471 (main)
+* https://github.com/thanos-io/thanos/issues/448 (store gateway OOM)
+
+## Summary
+
+This document describes the motivation and design of the `BucketFile` of the store component. 
+
+Our goals here are:
+
+- A) Reduce memory usage
+- B) Reuse chunks fetched from object storage
+
+## Motivation
+
+Currently, the store gateway loads all [index.cache.json](https://github.com/thanos-io/thanos/blob/865d5ec710b5651ead08a7dc73864a6f5dbeeaa9/pkg/block/index.go#L40)  files into memory on startup, which easily hits OOM situations. On the other hand, some indexes are rarely or never queried, so we don't have to keep these indexes in memory.
+
+The store gateway fetches chunks from object storage for each query request. but fetched chunks are not reused for next requests.
```

Comment:
If we will `mmap` everything and create sparse files then won't this technically mean that there is kind of an assumption that we have "infinite" RAM and address space? How will `BucketFile` ensure that we won't pass a certain threshold? What will we do in situations under memory pressure? I think that ATM having a common pool of RAM for responses serves us well as a capacity planning mechanism but this could potentially mean that someone will just ask for random data from a bucket that could be terabytes in size, and we will hold all of that in memory?

Reply:
>  ask for random data from a bucket that could be terabytes in size, and we will hold all of that in memory?    

will not hold all of that in memory.  
`BucketFile` will read small size (e.g. 1 MiB) data from `io.Reader` and write to a local file each time, until  `io.Reader` EOF. 

> there is kind of an assumption that we have "infinite" RAM and address space?

If we fetch many blocks (> 1000 blocks) from object storage,
it is possible that the size of all mmaps exceeds the virtual address space (e.g. 128TiB) for the store gateway process.
I think [Thanos Sharding](https://github.com/thanos-io/thanos/blob/master/docs/proposals/201909_thanos_sharding.md) is a good way to resolve this.

Reply:
Yeah but typically resilient systems gracefully shed their load, no? I would rather have Thanos Store unmmap some files instead of crashing in the middle of the night when someone queries a bunch of disparate data. I am fairly confident that we should probably have some kind of strategy on how to deal with this kind of situation instead of crashing and telling our users that maybe they should split up the Thanos Store nodes.

Reply:
Add a helper struct `MmapTracker` to manage the size of all mmaps.

The`MmapTracker` is a global obj for store gw. 

For each `BucketFile`:
1. call `MmapTracker.Allocate(size)` before `mmap` a file
2. call `MmapTracker.Free(size)` after `unmmap` a file

`MmapTracker.allocate` returns an error if the current size is greater than the configurable `MaxMmapSize`.

How about this way?  @bwplotka @GiedriusS

Reply:
Again, for design it's really enough to say `We will track mmaped file globally per store.`

And then knowing that there are important design questions to ask, e.g
* What if one request take all MaxMmapSIze?
* Who unmmaps file? Is it get unmmapped after each request?
* We fetch index and chunk pieces on demand but once we fetch all, are we mmapping the whole file? Potentially large one?